Layout: use CSS logical properties for margin top and bottom

ramonjd opened the pull request:
Depends on

- https://github.com/WordPress/gutenberg/pull/38883

## Description
The layout implementation doesn't take into account different writing modes, such as left-to-right, vertical and so on. 

With [CSS logical properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties), block and inline margins will apply to the appropriate side of an element depending on the direction of the document flow. 

This change to the layout margins ensures that margins will adhere to the logic of the current flow. 

Related PR: 
- https://github.com/WordPress/gutenberg/pull/38319

For example, margin-block-start (instead of margin-top) will manifest itself as a top margin for `writing-mode: horizontal-tb;` but a right margin for `writing-mode: vertical-rl;`.

#### With writing-mode: horizontal-tb;

<img width="700" alt="Screen Shot 2022-02-15 at 9 15 30 pm" src="https://user-images.githubusercontent.com/6458278/154042083-aa1ebf93-a115-423e-83d4-b6adaf044add.png">

#### With writing-mode: vertical-rl;

Trunk:

<img width="700" alt="Screen Shot 2022-02-15 at 9 18 07 pm" src="https://user-images.githubusercontent.com/6458278/154042101-4f2000cb-543c-453b-b208-7756b1d2b181.png">

This branch:

<img width="700" alt="Screen Shot 2022-02-15 at 9 18 45 pm" src="https://user-images.githubusercontent.com/6458278/154042104-2f1efca9-cb6f-45d6-b0f0-e06ac54a8d87.png">

## TODO
- [x] Update [the block rules in the WP_Theme_JSON class](https://github.com/wordpress/wordpress-develop/blob/2796abaf53c5a3b534c61ad0979fd25011174746/src/wp-includes/class-wp-theme-json.php#L827) via compat/6.0

## Why do this?

As mentioned over at https://github.com/WordPress/gutenberg/issues/38694#issuecomment-1039883033, this change will make it easier for theme authors to style block content using alternative `writing-mode`s.

## Testing Instructions

Test both in WordPress 5.8 and WordPress 5.9.

Open the editor and add a group block with several children. Here is some test code:

<details><summary>Test block code</summary>

```html
<!-- wp:group {"style":{"spacing":{"blockGap":"100px"}},"backgroundColor":"vivid-red","textColor":"white","layout":{"inherit":false}} -->
<div class="wp-block-group has-white-color has-vivid-red-background-color has-text-color has-background"><!-- wp:paragraph -->
<p>見へや今点タ将職び事確にみ率佳わ童能ぼずイ済番ラノ督扱動チサ浦交がわ二決シ広安3観売ノイ雇毎レヨロル間入いまもク質声ルイ以農閣りう。</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>事び急会ワ妹隠うーか立去確オア事限68戦ヲ案済水タセ帳容んげて化53危埼な委豪再三以けな。</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Test元読リれづご現45使ぼルき年視だレ大来かッドへ全額んぎゆ変78考ナ米去べふて出載ホ将秋おふ徹間供ノネアケ人購クムネ審花飛ふっ。Test</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>無のべ給転ルアカテ真姿ま初新チノミ問堀チノニイ問済ソサノ初重ムワモ総買ワ研型ぶふドけ済和ヨ正連きたど学6人フ厚供つスじト素考ムリユク八暮候むだ。</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```


</details>


Play around with the gap value, then publish the post and take a squiz at the frontend.

Check that the gap values between the group's child elements (margin top) appear as expected.

Using the browser inspector, add `writing-mode: vertical-rl;` to the group container div.

<img width="400" alt="Screen Shot 2022-02-15 at 9 38 29 pm" src="https://user-images.githubusercontent.com/6458278/154045151-89813a5c-c4d2-45a6-9797-77fa2a9ea844.png">

Check that the gap values between the group's child elements (logical margin top (right)) appear as expected.


| Former CSS property  | Equivalent logical property |
| ------------- | ------------- |
| margin-top  | margin-block-start |
| margin-bottom  | margin-block-end  |

Use a theme that opts into `blockGap` such as TwentyTwentyTwo.

Check that the gap between site blocks are uneffected.

For block themes, `.wp-site-blocks` should have an inline class of `wp-site-blocks > * + * { margin-block-start: var( --wp--style--block-gap );`

Run the tests!

```bash
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php

```

## Types of changes
CSS enhancement. Future compatibility.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

ramonjd committed:
Layout doesn't take into account different writing modes, such as left-to-right, vertical and so on. With CSS logical properties, block and inline margins will apply to the appropriate side depending on the direction of the document flow. This change to the layout margins ensures that margins will adhere to the logic of current flow. For example, margin-block-start (instead of margin-top) will manifest itself as a top margin for `writing-mode: horizontal-tb;` but a right margin for `writing-mode: vertical-rl;`.

fwazeter commented:
In my projects, I've been using an implementation of this in my own overrides of the Render Layout Support portion of the block supports API for months now across about a dozen or so production FSE sites in the wild and they've all worked flawlessly.

It's implementation makes sense because:
1) The semantics follow flex/grid properties, which are a must these days in CSS.
2) The caniuse for the non-shorthand properties are 96%, only really missing IE 11 which we don't support anyway.
3) It's global internationalization by default without the need for more stylesheets, hurray!

ramonjd committed:
Fixing CSS error where I mean to use margin-inline-end for margin-right. Other minor formatting
Taking a clumsy stab at compat files so we can load the global styles and settings.

ramonjd committed:
Updated tests and removed var_dumpy

ramonjd commented on lib/compat/wordpress-5.9/script-loader.php:
```diff
@@ -28,7 +28,7 @@ function gutenberg_enqueue_global_styles_assets() {
 	}
 
-	$stylesheet = wp_get_global_stylesheet();
+	$stylesheet = gutenberg_get_global_stylesheet();
```
I'm not sure whether I should be moving this entire file and [get-global-styles-and-settings.php](https://github.com/WordPress/gutenberg/pull/38816/files#diff-c09b194b1a3a092c11c1afc186045111e320e498cd34305b314774733d43bb2a) to `compat/6.0` or leaving them here.



The[ PR description](https://github.com/WordPress/gutenberg/pull/38745) for most recent changes to this file states that "The global styles stylesheet should be loaded using this logic, no matter the WordPress version", and a comment in [load.php](https://github.com/WordPress/gutenberg/blob/0b536eb460b5c424514be962df0941f78e2c5df5/lib/load.php) says:

> [script-loader.php]  Needs to be loaded before get-global-styles-and-settings.php

If we need to make similar changes such as the one this PR is attempting, then perhaps it's worth moving both files in their entirety to 6.0 (along with these changes)?

By that I mean

```diff
diff --git a/lib/load.php b/lib/load.php
index 18d541a969..a4babbfe25 100644
--- a/lib/load.php
+++ b/lib/load.php
@@ -73,8 +73,6 @@ require __DIR__ . '/compat/wordpress-5.9/register-global-styles-cpt.php';
 // to make sure we can use the check "function_exists( 'wp_get_global_styles' )".
 // If it loads after, that function will always be present at that point
 // and the global styles assets won't be loaded.
-require __DIR__ . '/compat/wordpress-5.9/script-loader.php';
-require __DIR__ . '/compat/wordpress-5.9/get-global-styles-and-settings.php';
 require __DIR__ . '/compat/wordpress-5.9/render-svg-filters.php';
 require __DIR__ . '/compat/wordpress-5.9/json-file-decode.php';
 require __DIR__ . '/compat/wordpress-5.9/translate-settings-using-i18n-schema.php';
@@ -96,6 +94,7 @@ require __DIR__ . '/compat/wordpress-5.9/default-theme-supports.php';
 require __DIR__ . '/compat/wordpress-5.9/class-wp-rest-global-styles-controller.php';
 require __DIR__ . '/compat/wordpress-5.9/rest-active-global-styles.php';
 require __DIR__ . '/compat/wordpress-5.9/move-theme-editor-menu-item.php';
+require __DIR__ . '/compat/wordpress-6.0/script-loader.php';
 require __DIR__ . '/compat/wordpress-6.0/get-global-styles-and-settings.php';
 require __DIR__ . '/compat/wordpress-6.0/post-lock.php';
 require __DIR__ . '/compat/wordpress-6.0/blocks.php';
```

I'm not very experienced in the compat wizardry so 🤷

ramonjd replied:
Once this is rebased on top of https://github.com/WordPress/gutenberg/pull/38883 my question will be answered :)
In short, yes.

ramonjd commented on lib/global-styles.php:
```diff
@@ -53,15 +53,15 @@ function_exists( 'gutenberg_is_edit_site_page' ) &&
 		$new_global_styles = array();
 
-		$css_variables = wp_get_global_stylesheet( array( 'variables' ) );
+		$css_variables = gutenberg_get_global_stylesheet( array( 'variables' ) );
```
I'm not entirely convinced that a rename here is necessary for the changes in this PR, but I thought it'd be safer to port them.

ramonjd replied:
Actually, they are :) See https://github.com/WordPress/gutenberg/pull/38883

ramonjd committed:
Removing reference to non-existent file after rebasing on top of #38883
Removing `get_stylesheet()` from WP_Theme_JSON_Gutenberg

ramonjd commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -42,4 +42,79 @@ public function get_patterns() {
+				$has_block_gap_support = _wp_array_get( $this->theme_json, array( 'settings', 'spacing', 'blockGap' ) ) !== null;
+				if ( $has_block_gap_support ) {
+					$block_rules .= '.wp-site-blocks > * { margin-block-start: 0; margin-block-end: 0; }';
+					$block_rules .= '.wp-site-blocks > * + * { margin-block-start: var( --wp--style--block-gap ); }';
```
This method was imported to redefine these styles.

andrewserong reviewed (commented):
Thanks for working on this @ramonjd, this looks like a good path forward for supporting different languages and writing modes in our layout supports 👍

I couldn't find any issues immediately with this PR applied, except that when I switched `writing-mode: vertical-rl` on within the front end of my site running TT2, the margins then behave differently to how I'd expect. While the paragraphs and spacing rendered correctly in a vertical layout, the horizontal margins that were aligning the block centred in the page, now cause the block to display at the left side of the page:

| Before applying writing mode | After applying writing mode |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/155064076-c940d12b-ffae-4da8-a382-3777ae9cda05.png) | ![image](https://user-images.githubusercontent.com/14988353/155064089-dc060cc3-270b-4799-8c6e-8d8fd9420b21.png) |

Did you encounter anything like that in your testing? I manually applied the CSS property in Dev Tools to the group block.

Update: I _think_ that's because of `margin-block-start` and `margin-block-end` being set to `0` in the vertical writing mode. If they're `auto` then I can get it centred again?

![image](https://user-images.githubusercontent.com/14988353/155064531-ea0ecab0-7e26-4cc0-8210-b8f63d0e7f35.png)

So I wonder if there's a limit to how many of the rules we want to be _logical_ (and in which circumstances), versus physical? E.g. in vertical writing mode, how much of the layout do we want it to automatically change?

ramonjd commented:
> So I wonder if there's a limit to how many of the rules we want to be logical (and in which circumstances), versus physical? E.g. in vertical writing mode, how much of the layout do we want it to automatically change?

Thanks for testing @andrewserong and for pointing out the inconsistency there.

This one was always going to be cutting edge, given that there's no immediate requirement for CSS logical properties and seeing as we don't officially support `writing-mode`.

 It was really meant as a stepping stone towards getting full support if we ever refactor layout, and also to set theme authors on an easier path to their own writing mode implementations.

> I think that's because of margin-block-start and margin-block-end being set to 0 in the vertical writing mode. If they're auto then I can get it centred again?

This makes sense to me. I didn't see this when testing, but I agree it's a bit strange. I'll have another look to try and get it working more intuitively.

Thanks again!

ramonjd committed:
Removing the CSS logical properties for styles that control horizontal page layout (left and right margins) so that any changes to the writing mode of blocks within wp-site-blocks and block containers still align themselves according to current horizontal rules.

kevin940726 reviewed (approved):
Just want to say that this makes sense to me, and the code (JS/CSS) looks good too!

ramonjd commented:
Thanks for testing @kevin940726 ! 


> I think that's because of margin-block-start and margin-block-end being set to 0 in the vertical writing mode. If they're auto then I can get it centred again?

@andrewserong The PR is now limited to swapping `margin-top` and `margin-bottom` for `margin-block-start` and `margin-block-end` respectively.

The right and left values I've left alone to preserve centered layouts even in alternative writing-modes.

![2022-03-10 12 20 56](https://user-images.githubusercontent.com/6458278/157568806-f0771e55-aab7-4854-9903-82e1fa6b85f2.gif)

I've rebased and tested the changes again.

andrewserong reviewed (approved):
> @andrewserong The PR is now limited to swapping margin-top and margin-bottom for margin-block-start and margin-block-end respectively.

Thanks for the ping @ramonjd! Yes, that's fixed it up for me so the centered styling is preserved now in testing:

![image](https://user-images.githubusercontent.com/14988353/157585447-81f337ee-57e7-4a93-9f38-8c7a7bc2cf75.png)

I've given this a re-test with a few different blocks, and tested in TT2 and looked for regressions in TT1 and TwentyTwenty themes, and couldn't find any issues 🙂 

This LGTM, and we can always look into follow-ups if needed for https://github.com/WordPress/gutenberg/pull/38319 down the track!
